domain/src/main/kotlin/com/krzykrucz/elesson/currentlesson/startlesson/domain/WorkflowImplementation.kt

```diff
@@ -1,35 +1,54 @@
-import com.krzykrucz.elesson.currentlesson.shared.mapSuccess
-import com.krzykrucz.elesson.currentlesson.startlesson.domain.StartLessonError.ClassRegistryUnavailable
-import com.krzykrucz.elesson.currentlesson.startlesson.domain.StartLessonError.NotScheduledLesson
+import com.krzykrucz.elesson.currentlesson.shared.sequence
 
 
 private fun ScheduledLesson.lessonIdentifier() =
-        LessonIdentifier(this.scheduledTime.toLocalDate(), this.lessonHourNumber, this.className)
-
-private fun ScheduledLesson.toCurrentLessonWithClass(classRegistry: ClassRegistry) =
-        StartedLesson(this.lessonIdentifier(), classRegistry, this.subject)
-
-
-fun startLesson(checkScheduledLesson: CheckScheduledLesson,
-                fetchClassRegistry: FetchClassRegistry): StartLesson = { teacher, attemptedStartTime ->
-    checkScheduledLesson(teacher, attemptedStartTime)
-        .mapError { NotScheduledLesson() }
-        .failIf({ scheduledLesson -> attemptedStartTime.isBefore(scheduledLesson.scheduledTime) }, NotScheduledLesson())
-        .failIf({ scheduledLesson -> attemptedStartTime.isAfter(scheduledLesson.scheduledTime.plusMinutes(44)) }, NotScheduledLesson())
-        // ^ TODO maybe enclose these 2 mapping in the above and introduce another internal error type (lesson started too early or so)
-        // ^ TODO extract a separate function with invariant
-        .flatMapAsyncSuccess { scheduledLesson ->
-            fetchClassRegistry(scheduledLesson.className)
-                .mapSuccess(scheduledLesson::toCurrentLessonWithClass)
-                .mapError { ClassRegistryUnavailable() }
-        }
+    LessonIdentifier(this.scheduledTime.toLocalDate(), this.lessonHourNumber, this.className)
+
+private fun ScheduledLesson.toCurrentLessonWithClass(
+    classRegistryOrError: Either<StartLessonError, ClassRegistry>
+): Either<StartLessonError, StartedLesson> =
+    classRegistryOrError
+        .map { classRegistry -> StartedLesson(this.lessonIdentifier(), classRegistry, this.subject) }
+
+
+fun startLesson(
+    fetchScheduledLesson: FetchScheduledLesson,
+    validateLessonStartTime: ValidateLessonStartTime,
+    fetchClassRegistry: FetchClassRegistry
+): StartLesson = { teacher, attemptedStartTime ->
+    IO.fx {
+        val (scheduledLessonOrError) =
+            fetchScheduledLesson(teacher, attemptedStartTime)
+
+        val validatedScheduledLesson = scheduledLessonOrError
+            .flatMap { validateLessonStartTime(it, attemptedStartTime) }
+
+        val (startedLessonOrError) = validatedScheduledLesson
+            .map { scheduledLesson ->
+                fetchClassRegistry(scheduledLesson.className)
+                    .map(scheduledLesson::toCurrentLessonWithClass)
+            }.sequence()
+
+        startedLessonOrError.flatten()
+
+    }
 }
 
+
+fun validateStartTime(): ValidateLessonStartTime = { lesson, time ->
+    when {
+        time.isBefore(lesson.scheduledTime) -> Either.left(StartLessonError.NotScheduledLesson())
```

Comment:
please reinstate the following TODO

// TODO introduce another internal error type (lesson started too early or so)

Reply:
I will commit that to the PR before merge :)